temporalio/worker/_workflow_instance.py

```diff
@@ -240,6 +267,26 @@ def __init__(self, det: WorkflowInstanceDetails) -> None:
         self._queries = dict(self._defn.queries)
         self._updates = dict(self._defn.updates)
 
+        # We record in-progress signals and updates in order to support waiting for handlers to
+        # finish, and issuing warnings when the workflow exits with unfinished handlers. Since
+        # signals lack a unique per-invocation identifier, we introduce a sequence number for the
+        # purpose.
+        self._handled_signals_seq = 0
+        self._in_progress_signals: dict[
+            int,
+            Tuple[
+                temporalio.bridge.proto.workflow_activation.SignalWorkflow,
```

Comment:
While technically storing the activation for the life of the handler has memory consequences, we already suffer them since we use nested functions I think. But I do think it'd be clearer having some kind of `InProgressHandler` dataclass with just what you need as opposed to a tuple.

Reply:
I agree it'd be nice to have some dataclass here. Especially if we add support for the more general API I mentioned before where you can get a list of all running handlers and decide what you want to do with them.